src/key_info_managers/mod.rs

```diff
@@ -146,16 +149,155 @@ pub trait ManageKeyInfo {
     fn exists(&self, key_triple: &KeyTriple) -> Result<bool, String>;
 }
 
-// "+ Send + Sync" is needed for things like:
-// ```
-// let store_handle = self.key_info_store.read().expect("Key store lock poisoned");
-// let _ = store_handle.list_keys(&app_name, ProviderID::Pkcs11)?;
-// ```
-// to work because `store_handle` is a RWLockReadGuard of (dyn ManageKeyInfo + Send + Sync +
-// 'static). Did not work with only `impl dyn ManageKeyInfo` below. Maybe there is a way to
-// implement automagically the same methods on (dyn ManageKeyInfo + Send + Sync + 'static)
-// if they are implemented on dyn ManageKeyInfo but not sure.
-impl dyn ManageKeyInfo + Send + Sync {
+/// KeyInfoManager client structure that bridged between the KIM and the providers that need
+/// to use it.
+#[derive(Derivative)]
+#[derivative(Debug)]
+pub struct KeyInfoManagerClient {
+    provider_id: ProviderID,
+    #[derivative(Debug = "ignore")]
+    key_info_manager_impl: Arc<RwLock<dyn ManageKeyInfo + Send + Sync>>,
+}
+
+impl KeyInfoManagerClient {
+    /// Create a new KeyInfoManager client
+    pub fn new(
+        provider_id: ProviderID,
+        key_info_manager_impl: Arc<RwLock<dyn ManageKeyInfo + Send + Sync>>,
+    ) -> Self {
+        KeyInfoManagerClient {
+            provider_id,
+            key_info_manager_impl,
+        }
+    }
+
+    /// Get the KeyTriple representing a key.
+    pub fn get_key_triple(&self, app_name: ApplicationName, key_name: String) -> KeyTriple {
+        KeyTriple::new(app_name, self.provider_id, key_name)
+    }
+
+    /// Get a key ID of any type implementing `Deserialize`
+    pub fn get_key_id<T: serde::de::DeserializeOwned>(
+        &self,
+        key_triple: &KeyTriple,
+    ) -> parsec_interface::requests::Result<T> {
+        let key_info_manager_impl = self
+            .key_info_manager_impl
+            .read()
+            .expect("Key Info Manager lock poisoned");
+        let key_info = match key_info_manager_impl.get(key_triple) {
+            Ok(Some(key_info)) => key_info,
+            Ok(None) => return Err(ResponseStatus::PsaErrorDoesNotExist),
+            Err(string) => return Err(to_response_status(string)),
+        };
+        // the ID is cloned below because `deserialize` links the lifetime of the
+        // input to that of the output, and the input is tied down to the RWLockReadGuard
+        Ok(bincode::deserialize(&key_info.id)?)
+    }
+
+    /// Get the `Attributes` of a key
+    pub fn get_key_attributes(
```

Comment:
Well, I was looking at that - it would be possible, and in some places we do need both, but I think in most cases only one is needed. I just didn't like returning a tuple - I could make a "generic" `KeyInfo`, but that'd be a bit annoying

Reply:
Yes that is a very good point actually... Why don't we just put the `T` inside the `KeyInfo` structure??

Reply:
I think it's possible, but making a type like this generic would be a bit rough on the interfaces that we need to change, so I'd leave that for another issue/PR (I did try it out locally and it's not that straight-forward). I think the question here should be whether we return both values as a tuple or not.

Reply:
Ah yes, I think it's better to have clean interfaces. My preference would be to return both values as a tuple here, and ignore what we don't need when we call the method.

Reply:
Bumped into a problem - if we only want the `attributes`, we still have to make the ID type explicit

Reply:
.. annoying... Yes I guess this is fine like that then